Introduce discovery_service and automatically run an SSM Document on discovered EC2 nodes

lxea opened the pull request:
Make use of the EC2 fetcher to run a specified SSM document on the discovered EC2 instances 

This plumbs changes from some supporting PRs (#13886 , #13787) so depends on both of these.

It also adds moves configuration over to `discovery_service` instead of SSH.

Configuration now looks like :
```yaml
discovery_service:
  enabled: true
  aws:
   - types: ["ec2"]
     regions: ["eu-central-1"]
     tags:
       "discover_teleport": "yes"
     install:
       join_params:
         token_name:  iam-invite-token
     ssm:
       document_name: "test-installer-document"
```

TODO: a unit test exists but needs improvements

Commit:
Add initial version of installer

r0mant commented on api/types/events/events.proto:
```diff
@@ -2505,3 +2506,21 @@ message UpgradeWindowStartUpdate {
         [ (gogoproto.nullable) = false, (gogoproto.embed) = true, (gogoproto.jsontag) = "" ];
 }
+
+message EC2DiscoveryScriptRun {
```
I think this is too specific, how about `SSMRun`? If we want to differentiate further, we could add "type" field or something in future.

Also, missing godoc.

r0mant commented on api/types/events/events.proto:
```diff
@@ -2505,3 +2506,21 @@ message UpgradeWindowStartUpdate {
+    string CommandID = 2 [ (gogoproto.jsontag) = "command_id" ];
+
+    // InstanceID is the id of the EC2 instance the command was run on.
+    string InstanceID = 3 [ (gogoproto.jsontag) = "instance_id" ];
```
Is there any more information we can/should add? E.g. account ID, region, etc?

r0mant commented on lib/events/api.go:
```diff
@@ -537,11 +537,14 @@ const (
 
+	// EC2DiscoveryInstallScriptEvent is emitted when a run of an
+	// install script completes on a discovered EC2 node
+	EC2DiscoveryInstallScriptEvent = "discovery.ec2.install_script"
```
`ssm.run` as I suggested above.

r0mant commented on lib/events/codes.go:
```diff
@@ -329,6 +329,9 @@ const (
 	// UpgradeWindowStartUpdatedCode is the edit code of UpgradeWindowStartUpdateEvent.
 	UpgradeWindowStartUpdatedCode = "TUW01I"
 
+	DiscoveryScriptEC2SuccessCode = "TDS00I"
+	DiscoveryScriptEC2FailCode    = "TDS00W"
```
Godocs.

r0mant commented on lib/srv/regular/sshserver.go:
```diff
@@ -1019,6 +1027,49 @@ func (s *Server) getServerResource() (types.Resource, error) {
+func (s *Server) filterExistingNodes(instances server.EC2Instances) (server.EC2Instances, error) {
+	var filtered []*ec2.Instance
+	for _, inst := range instances.Instances {
+		instID := fmt.Sprintf("%s-%s",
+			instances.AccountID,
+			aws.StringValue(inst.InstanceId))
```
I think we should be filtering by label, not node name which can get overridden.

The install script should make sure that the joining node includes AWS instance ID in a label e.g. `teleport.dev/instance-id: i-xyz123` and then we should filter by label here.

r0mant commented on lib/srv/regular/sshserver.go:
```diff
@@ -1019,6 +1027,49 @@ func (s *Server) getServerResource() (types.Resource, error) {
+
+func (s *Server) handleEC2Discovery() {
+	go s.cloudWatcher.Run()
+	for inst := range s.cloudWatcher.Instances {
```
Instead of "for", this should select on the channel and ctx.Done.

r0mant commented on lib/srv/regular/sshserver.go:
```diff
@@ -1019,6 +1027,49 @@ func (s *Server) getServerResource() (types.Resource, error) {
+			log.Errorf("Error filtering existing nodes: %s", err)
+		}
+		installer := server.NewInstallation(client, inst.Instances, inst.Parameters)
+		results, err := installer.DoInstall(inst.Document)
```
Does this execute synchronously or asynchronously?

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,122 @@
+	"github.com/gravitational/trace"
+)
+
+type Installation struct {
```
I would just call it `SSMInstaller` and maybe move into lib/cloud too.

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,122 @@
+	params    map[string][]*string
+}
+
+func NewInstallation(client ssmiface.SSMAPI, instances []*ec2.Instance, params map[string]string) *Installation {
```
`NewSSMInstaller`. Also, convert parameters to a proper config struct.

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,122 @@
+	return resultCmds, nil
+}
+
+func (i *Installation) DoInstall(document string) ([]*events.EC2DiscoveryScriptRun, error) {
```
```suggestion
func (i *Installation) Run(document string) ([]*events.EC2DiscoveryScriptRun, error) {
```

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,122 @@
+
+var ErrCommandInProgress = errors.New("command in progress")
+
+func (i *Installation) checkCommands(commandID *string) ([]*events.EC2DiscoveryScriptRun, error) {
```
What is "check commands" and why does it return audit events? It's a little confusing interface.

I would just initialize SSMInstaller with the audit log interface and have it emit events.

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,122 @@
+	}
+
+	return &Installation{
+		instances: ids,
```
Why is the installer initialized with instances? Wouldn't it make more sense to accept instance as a part of the install/run method?

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,122 @@
+
+	commandID := output.Command.CommandId
+	for {
+		<-i.rechecker.C
```
This needs some cancellation support.

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,165 @@
+
+var ErrCommandInProgress = errors.New("command in progress")
+
+func (i *SSMInstaller) checkCommands(commandID *string) error {
```
Nit: I'd move this method after `Run()` so it's easier to read the flow.

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,165 @@
+		}
+		status := aws.StringValue(cmdOut.Status)
+		if status == ssm.CommandStatusInProgress {
+			return trace.Wrap(ErrCommandInProgress)
```
I think you can just return `trace.AlreadyExists` error here, then we don't need an extra error type.

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,165 @@
+	}
+}
+
+var ErrCommandInProgress = errors.New("command in progress")
```
As mentioned below, you can prob. just use `trace.AlreadyExists` which fits thematically, and remove this one.

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,165 @@
+}
+
+// Run executes the SSM document and then blocks until the command has completed.
+func (i *SSMInstaller) Run(document string) error {
```
I think this method should take a context.
```suggestion
func (i *SSMInstaller) Run(ctx context.Context, document string) error {
```

r0mant commented on lib/srv/regular/sshserver.go:
```diff
@@ -1047,6 +1059,60 @@ func (s *Server) getServerResource() (types.Resource, error) {
+			if err != nil {
+				log.Errorf("Error filtering existing nodes: %s", err)
+			}
+			installer := server.NewSSMInstaller(
+				server.SSMInstallerConfig{
+					SSM:       client,
+					Instances: filteredInstances.Instances,
+					Params:    filteredInstances.Parameters,
+					Emitter:   s,
+					Ctx:       s.ctx,
+					Region:    filteredInstances.Region,
+					AccountID: filteredInstances.AccountID,
+				})
+			if err := installer.Run(filteredInstances.Document); err != nil {
```
I think I asked before, but it's a little confusing interface that SSM installer is initialized with instances, context, etc. Those should be parameters to Run.

Can we instantiate SSMInstaller once (make it a part of Server for example) with SSM client and Emitter and make other fields parameters to Run method?

```go
type SSMInstallerConfig struct {
  SSM ...
  Emitter ...
}

type SSMRunRequest struct {
  Instances ...
  Params ...
  Region string
  AccountID string
}

func (i *ssmInstaller) Run(ctx context.Context, req SSMRunRequest) error {
  ...
}
```

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,165 @@
+			if errors.Is(err, ErrCommandInProgress) {
+				continue
+			}
+			return err
```
```suggestion
			return trace.Wrap(err)
```

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,165 @@
+	return &SSMInstaller{
+		instances:    ids,
+		SSM:          cfg.SSM,
+		recheckTimer: time.NewTicker(time.Second * 30),
```
If you implement my suggestion above with structs reorg, this ticket will just move to Run method.

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,165 @@
+			Status:     status,
+		}
+
+		if err := i.emitter.EmitAuditEvent(i.ctx, &event); err != nil {
```
The `checkCommands` function can run multiple times for the same set of servers, right? If e.g. some are still executing. Doesn't this mean that this will emit multiple audit events for the same server i.e. we'll get duplicates?

codingllama commented:
It looks like you are still iterating here, so I'll unassign. Feel free to add me back when you're good to go.

Commit:
Resolve comments

- Use aws waiters when checking commands
- Use SSMRunRequest rather than passing instances
- General comments

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,150 @@
+		return trace.Wrap(err)
+	}
+
+	var g errgroup.Group
+	for _, inst := range ids {
+		g.Go(func() error {
+			return trace.Wrap(i.checkCommand(ctx, req, output.Command.CommandId, inst))
+		})
+	}
+	return trace.Wrap(g.Wait())
```
We should probably limit the parallelism here to something reasonable like 10, in case many nodes are joining at once. Errgroup has a SetLimit method.

tigrato replied:
And maybe call it with a cancelable context

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,150 @@
+		CommandId:  commandID,
+		InstanceId: instanceID,
+	})
+	if err := checkAWSWaitErr(err); err != nil {
+		return trace.Wrap(err)
+	}
```
What does this error indicate?

r0mant reviewed (commented):
@tigrato @capnspacehook Can we get reviews here please?

tigrato commented on lib/auth/api.go:
```diff
@@ -142,6 +142,9 @@ type ReadNodeAccessPoint interface {
 	GetNetworkRestrictions(ctx context.Context) (types.NetworkRestrictions, error)
+
+	// GetNodes returns a list of registered servers for this cluster.
+	GetNodes(ctx context.Context, namespace string) ([]types.Server, error)
```
Should it be `GetServers`  to keep the naming logic?

tigrato commented on lib/srv/ctx.go:
```diff
@@ -101,6 +101,9 @@ type AccessPoint interface {
 	GetCertAuthorities(ctx context.Context, caType types.CertAuthType, loadKeys bool, opts ...services.MarshalOption) ([]types.CertAuthority, error)
+
+	// GetNodes returns a list of registered servers for this cluster.
+	GetNodes(ctx context.Context, namespace string) ([]types.Server, error)
```
same logic as before

tigrato commented on lib/srv/regular/sshserver.go:
```diff
@@ -1013,6 +1025,70 @@ func (s *Server) getServerResource() (types.Resource, error) {
 	return s.getServerInfo(), nil
 }
 
+func (s *Server) filterExistingNodes(instances *server.EC2Instances) (*server.EC2Instances, error) {
```
can we keep a cached version of this data somehow?
For large ec2 fleets, `GetNodes` might extract a lot of info (including nodes that weren't discovered by ec2 discovery) from AuthServer and we can cache everything locally.

One way:
On load, the agent loads from Auth server the list of discovered servers (filter by tag so we do not receive other nodes) and stores them locally. Each time a new machine is discovered(not in cache) you can just ping AuthServer to see if no one else discovered that machine already and you update your local cache. If the machine was not discovered yet, you just run SSM against it

r0mant replied:
I talked to @lxea about this today, we already use AccessPoint here which uses local cache rather than calling directly Auth server so I think we should be good.

tigrato commented on lib/srv/regular/sshserver.go:
```diff
@@ -1013,6 +1025,70 @@ func (s *Server) getServerResource() (types.Resource, error) {
+		return trace.Wrap(err)
+	}
+	if len(filteredInstances.Instances) == 0 {
+		return trace.NotFound("all fetched nodes already registered")
```
Should we return an error for normal behavior?
I would say that we can return the number of enrolled nodes, and if it's 0 we just log a debug line saying the node has nothing else to do for now...

tigrato commented on lib/srv/regular/sshserver.go:
```diff
@@ -1013,6 +1025,70 @@ func (s *Server) getServerResource() (types.Resource, error) {
+	}
+
+	installer := server.NewSSMInstaller(server.SSMInstallerConfig{Emitter: s})
+	req := server.SSMRunRequest{
```
Can we trigger multiple times the SSM for the same machine? 
If for some reason the cluster/ssm is slow and the discovery runs multiple times before the SSM completes, can we end up with multiple SSM executions for the same machine?

SSM executions have costs, right?

r0mant replied:
@lxea and I discussed this today, we're going to use auth server's semaphore to make sure only 1 agent runs SSM on the instance.

tigrato commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+// Run executes the SSM document and then blocks until the command has completed.
+func (i *SSMInstaller) Run(ctx context.Context, req SSMRunRequest) error {
```
check my comment above about multiple runs for the same machine, but if possible should we just store the ids of the ec2 machines which we run SSM against but didn't finish yet?

tigrato commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,150 @@
+	return trace.Wrap(g.Wait())
+}
+
+func checkAWSWaitErr(err error) error {
```
add a description about what this does and see if the function should have a different name to be more meaningful

tigrato commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+func checkAWSWaitErr(err error) error {
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok &&
+			aerr.Code() == request.WaiterResourceNotReadyErrorCode {
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+	return nil
```
```suggestion
        var aErr awserr.Error
	if errors.As(err, &aErr) && aerr.Code() == request.WaiterResourceNotReadyErrorCode {
		return nil
	}
	return trace.Wrap(err)
```

tigrato commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,150 @@
+	return nil
+}
+
+func (i *SSMInstaller) checkCommand(ctx context.Context, req SSMRunRequest, commandID, instanceID *string) error {
```
can you use another var different than `i`? I thought we were in a loop.

capnspacehook commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,150 @@
+	var g errgroup.Group
+	for _, inst := range ids {
+		g.Go(func() error {
+			return trace.Wrap(i.checkCommand(ctx, req, output.Command.CommandId, inst))
```
`inst` loop variable is being captured here, should create a new variable to ensure `inst`'s value will not change while in `i.checkCommand`.

See https://golang.org/doc/go_faq.html#closures_and_goroutines

capnspacehook commented on lib/srv/server/ssm_install_test.go:
```diff
@@ -0,0 +1,152 @@
+		// todo(amk): test that incomplete commands eventually return
+		// an event once completed
+	} {
+		inst := NewSSMInstaller(tc.conf)
```
run each test case as a subtest so we can easily triage test failures?

Commit:
Resolve comments, (rebase) pass scriptname parameter

This resolves comments regarding running on multiple ec2 instances at
once by adding state to the instances cache to check if the instance
is known about and how far into installation it is

Commit:
Revert cache

r0mant reviewed (commented):
@lxea Can you rebase this PR off of the latest? Right now it also includes the multiple installers changes.

r0mant commented on api/types/constants.go:
```diff
@@ -345,6 +345,15 @@ const (
 	// created from the Kubernetes Operator.
 	OriginKubernetes = "kubernetes"
+
+	// AWSAccountIDLabel is used to identify nodes by AWS account ID
+	// found via automatic discovery, to avoid re-running installation
+	// commands on the node.
+	AWSAccountIDLabel = TeleportNamespace + "/account-id"
+	// AWSInstanceIDLabel is used to identify nodes by EC2 instance ID
+	// found via automatic discovery, to avoid re-running installation
+	// commands on the node.
+	AWSInstanceIDLabel = TeleportNamespace + "/instance-id"
```
I've been thinking, when we add auto-enrollment on other clouds (Azure, GCP), how do we differentiate which cloud the node originated from? These labels are pretty generic. How about we add another label, e.g. `teleport.dev/cloud` and set it to `aws` for auto-discovered EC2 instances?

r0mant replied:
The script isn't really cloud-dependent though, not sure we need different scripts just to pass an extra label.

It may be ok as-is though, this issue can probably be solved separately by adding support for cloud nodes to report their own metadata in heartbeats, then we can get rid of setting these labels in the script explicitly also. Let's keep it for now.

r0mant commented on api/types/events/events.proto:
```diff
@@ -2568,3 +2569,28 @@ message SessionRecordingAccess {
 }
+
+// SSMRun is emitted after an SSM document completes execution.
+message SSMRun {
```
Don't forget to update webapps to add support for this new event. Here's example: https://github.com/gravitational/webapps/pull/1100.

r0mant commented on lib/cache/cache.go:
```diff
@@ -229,6 +229,7 @@ func ForNode(cfg Config) Config {
 		{Kind: types.KindClusterAuthPreference},
 		{Kind: types.KindSessionRecordingConfig},
 		{Kind: types.KindRole},
+		{Kind: types.KindNode},
```
@rosstimothy @fspmarshall @espadolini This is going to cause problems with performance and/or network utilization in large clusters, right? A node that runs EC2 auto-discovery needs to query other nodes in the cluster to understand which nodes have already joined.

@lxea Can we make it so this is included in watch configuration only if the node runs auto-discovery? Otherwise if each node will watch other nodes is going to destroy large clusters.

r0mant commented on lib/srv/regular/sshserver.go:
```diff
@@ -1026,6 +1043,73 @@ func (s *Server) getServerResource() (types.Resource, error) {
+}
+
+func (s *Server) handleEC2Discovery() {
+
```
```suggestion
```

Commit:
Dont cache on non discovery nodes

Commit:
Resolve some comments

r0mant commented on lib/cache/cache.go:
```diff
@@ -235,6 +235,11 @@ func ForNode(cfg Config) Config {
 		{Kind: types.KindNetworkRestrictions},
 	}
+
+	if cfg.DiscoveryAgent {
```
They can deploy one or multiple, for example each discovering a subset of the nodes.

Yes, I did consider making it a separate service when working on database discovery but it felt like having a database agent run discovery for the databases it will proxy makes sense so we took the same approach for SSH. I think it's ok for now, but there's definitely merit to this idea, we'll probably revisit in the course of working on Teleport Discover.

r0mant replied:
Yes, exactly - the reason agents run discovery is to support Cloud. Still in theory would make sense to have `discovery_service` in agents which would support discovery for all types of resources we have, that would be quite significant refactoring though which is why I think we should do it when working on "day 2" Teleport Discover initiative (which will focus on auto-discovery).

r0mant replied:
We want to get this in 10.2 so it's a little too late for a redesign like this IMO.

r0mant commented on lib/cache/cache.go:
```diff
@@ -235,6 +235,11 @@ func ForNode(cfg Config) Config {
 	}
+
+	if cfg.DiscoveryAgent {
+		cfg.Watches = append(cfg.Watches, types.WatchKind{Kind: types.KindNode})
```
How about we make a separate `DiscoveryAccessPoint` interface that will have `GetNodes` method and will only get initialized in the server via the server option ([here](https://github.com/gravitational/teleport/blob/master/lib/service/service.go#L2257)) if discovery is enabled. Then it can return a proper error message if a node that does not have discovery enabled tries to fetch the nodes. Then we should also probably do separate `ForDiscovery` cache setup above rather than adding it to `ForNode`.

fspmarshall reviewed (commented):
I haven't been following the development of this feature, so I'm probably lacking some context.   I have a few comments/concerns tho:

Making this part of ssh service forces us to continue to grant `RoleNode` the ability to read every resource that it might require in order to complete this work.  Nodes already have more permissions than they really should, and we're actively working on trying to reduce their permission scope (e.g. nodes currently can read/write the node resource, but we want to remove that once we've been on the new heartbeat mechanism for at least one major version, this feature as implemented would prevent that).  Peripheral services should not have capabilities that allow them to manage/modify resources/servers other than the ones they are directly controlling access to.

Additionally, replicating/caching the entire node set represents the majority of all network load in large clusters.  Requiring another server to replicate the node resource (or any resource that scales with cluster size) should be a weapon of last resort, only if the feature in question absolutely cannot be implemented to run on the auth or proxy server.

It was suggested to me that maybe the reason this feature isn't just being added to auth is because of cloud deployments necessitating that auth/proxy live in a different cloud account than the one to which nodes are being deployed.  Assuming that's the case, and the root problem being attacked here is that duties which would normally be auth server duties must be performed by a peripheral agent in cloud deployments, then maybe we can try to solve that problem in the general case?  Maybe we could have some kind of `cloud_agent`, or maybe something more specialized like `aws_agent`/`gcp_agent`/etc.  Some service that folks can deploy one or two of to each cloud env, which could be a general-purpose solution to this class of problem? I think we're gonna hit some serious scaling issues if we don't decide on some kind of general solution so that future features like this can share a single cache.

r0mant commented:
@fspmarshall Yes, this has to run in an agent to support Cloud. Not every node will need to replicates all the nodes, only those that run auto-discovery, and we did talk about introducing a separate `discovery_service` as a general solution to all our auto-discovery needs which I agree would be cleaner.

@lxea Let's see how difficult it would be to factor this out into a separate service. Should be pretty straightforward, you can look at [database service](https://github.com/gravitational/teleport/blob/master/lib/service/db.go#L48) as an example. And I'll chat with the product team to see if we want to intro new `discovery_service` config section or keep it as part of `ssh_service` for now.

Commit:
Move discovery out to its own serviec

Commit:
Add a `discovery_service` section

Commit:
Fix messed up conflict merge

Commit:
Make starting a standalone discovery agent work

r0mant commented on lib/auth/api.go:
```diff
@@ -663,6 +663,67 @@ type WindowsDesktopAccessPoint interface {
+// used by a teleport.ComponentDiscovery.
+//
+// NOTE: This interface must match the resources replicated in cache.ForDiscovery.
+type ReadDiscoveryAccessPoint interface {
```
Does discovery service actually need all of these? Can we keep only those that it will actually use?

r0mant commented on lib/auth/permissions.go:
```diff
@@ -663,6 +663,29 @@ func definitionForBuiltinRole(clusterName string, recConfig types.SessionRecordi
+					Namespaces:           []string{types.Wildcard},
+					WindowsDesktopLabels: types.Labels{types.Wildcard: []string{types.Wildcard}},
+					Rules: []types.Rule{
+						types.NewRule(types.KindEvent, services.RW()),
+						types.NewRule(types.KindCertAuthority, services.ReadNoSecrets()),
+						types.NewRule(types.KindClusterName, services.RO()),
+						types.NewRule(types.KindClusterAuditConfig, services.RO()),
+						types.NewRule(types.KindClusterNetworkingConfig, services.RO()),
+						types.NewRule(types.KindSessionRecordingConfig, services.RO()),
+						types.NewRule(types.KindClusterAuthPreference, services.RO()),
+						types.NewRule(types.KindUser, services.RO()),
+						types.NewRule(types.KindRole, services.RO()),
+						types.NewRule(types.KindNamespace, services.RO()),
+						types.NewRule(types.KindLock, services.RO()),
+						types.NewRule(types.KindNode, services.RO()),
```
Same as above, I think we should trim these to the bare minimum the service needs. I may be mistaken but off the top of my head I'm not sure why it would need: audit config, networking config, session recording config, auth preference, user, role, lock.

r0mant commented on lib/cache/cache.go:
```diff
@@ -323,6 +324,26 @@ func ForWindowsDesktop(cfg Config) Config {
+		{Kind: types.KindRole},
+		{Kind: types.KindProxy},
+		{Kind: types.KindNamespace, Name: apidefaults.Namespace},
+		//		{Kind: types.KindNode},
```
Why is this commented out?

r0mant commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,172 @@
+	ec2Installer *server.SSMInstaller
+}
+
+func New(
+	ctx context.Context,
+	cfg *Config) (*Server, error) {
```
```suggestion
func New(ctx context.Context, cfg *Config) (*Server, error) {
```

r0mant commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,172 @@
+	for {
+		select {
+		case instances := <-s.cloudWatcher.InstancesC:
+			s.log.Debugln("EC2 instances discovered, starting installation")
```
Can we log discovered instances for debugging purposes?

r0mant reviewed (commented):
@lxea Nice job on the discovery service!

Have you tested scenarios where you can run it as a standalone service, as well as alongside ssh_service?

Commit:
Resolve comments

zmb3 commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -3663,3 +3664,31 @@ message SessionRecordingAccess {
   ];
 }
+
+// SSMRun is emitted after an SSM document completes execution.
```
```suggestion
// SSMRun is emitted after an AWS SSM document completes execution.
```

Extra context to indicate that this is AWS-specific will help for those who don't know that SSM is an AWS feature.

zmb3 commented on api/types/system_role.go:
```diff
@@ -68,6 +68,8 @@ const (
 	// is "primary".
 	RoleInstance SystemRole = "Instance"
+	// RoleDiscovery is a role for discovery nodes in the cluster
+	RoleDiscovery SystemRole = "Discovery"
```
Should `IsServer(RoleDiscovery)` return true or false?

zmb3 commented on lib/auth/api.go:
```diff
@@ -663,6 +663,43 @@ type WindowsDesktopAccessPoint interface {
 	accessPoint
 }
 
+// ReadDiscoveryAccessPoint is a read only API interface implemented by a certificate authority (CA) to be
```
What does "implemented by a certificate authority (CA)" mean here?

Discovery has nothing to do with issuing certificates, right?

zmb3 commented on lib/auth/permissions.go:
```diff
@@ -663,6 +663,22 @@ func definitionForBuiltinRole(clusterName string, recConfig types.SessionRecordi
+			types.RoleSpecV5{
+				Allow: types.RoleConditions{
+					Namespaces:           []string{types.Wildcard},
+					WindowsDesktopLabels: types.Labels{types.Wildcard: []string{types.Wildcard}},
```
You shouldn't need access to Windows Desktops in order to discover EC2 nodes.

zmb3 commented on lib/config/configuration.go:
```diff
@@ -2146,7 +2153,8 @@ func validateRoles(roles string) error {
 			defaults.RoleDatabase,
-			defaults.RoleWindowsDesktop:
+			defaults.RoleWindowsDesktop,
+			defaults.RoleDiscovery:
```
Is this necessary here? Would we ever expect a user to run `teleport start --roles=discovery` to run a standalone discovery service, or is discovery something that teleport should decide to enable based on the config?

Do we need to be able to generate join tokens for the discovery service, or does it always piggy back on another service?

zmb3 commented on lib/config/fileconf.go:
```diff
@@ -85,6 +85,10 @@ type FileConfig struct {
 	// Tracing is the "tracing_service" section in Teleport configuration file
 	Tracing TracingService `yaml:"tracing_service,omitempty"`
+
+	// Discoveryis the "discovery_service" section in the Teleport
```
```suggestion
	// Discovery is the "discovery_service" section in the Teleport
```

zmb3 commented on lib/config/fileconf.go:
```diff
@@ -1108,6 +1109,13 @@ func (ssh *SSH) X11ServerConfig() (*x11.ServerConfig, error) {
 	return cfg, nil
 }
 
+type Discovery struct {
```
Don't forget to update the config reference to include the new YAML.

r0mant replied:
```suggestion
// Discovery represents a discovery_service section in the config file.
type Discovery struct {
```

zmb3 commented on lib/defaults/defaults.go:
```diff
@@ -383,6 +383,9 @@ var (
 	// WindowsDesktopQueueSize is windows_desktop service watch queue size.
 	WindowsDesktopQueueSize = 128
 
+	// DIscoveryQueueSize is discovery service queue size.
```
```suggestion
	// DiscoveryQueueSize is discovery service queue size.
```

zmb3 commented on lib/events/dynamic.go:
```diff
@@ -249,8 +249,11 @@ func FromEventFields(fields EventFields) (events.AuditEvent, error) {
 		e = &events.SFTP{}
 	case UpgradeWindowStartUpdateEvent:
 		e = &events.UpgradeWindowStartUpdate{}
+
```
Remove arbitrary newline?

zmb3 commented on lib/service/discovery.go:
```diff
@@ -0,0 +1,121 @@
+	log := process.log.WithField(trace.Component, teleport.Component(
+		teleport.ComponentDiscovery, process.id))
+
+	conn, err := process.waitForConnector(DiscoveryIdentityEvent, log)
```
By making a new `DiscoveryIdentityEvent`, aren't we requiring node join tokens to have a new role for the discovery service? If so, this is going to break a lot of the existing docs.

Should discovery service instead try to use the identity of the SSH service?

(Honest question - I have a similar challenge with the uploader service)

r0mant replied:
@zmb3 The idea is to have a separate discovery service which you can run alongside other services (app, db, ssh) or separately. Reason is discovery service is a standalone component that just discovers resources and registers them in a cluster, with its own set of permissions and caches that it needs, etc. I had this among the topics for the today's product sync.

For the uploader service it seem that it can't really run as standalone and should be an internal service so the use-case is a little different.

zmb3 replied:
There's lots of places in the docs where we instruct users how to add a node to a cluster. We'll want to make sure to mention the new discovery role, otherwise people who enable discovery service in the config without ensuring their host certs have the discovery role will run into errors and discovery won't work.

zmb3 commented on lib/service/discovery.go:
```diff
@@ -0,0 +1,121 @@
+
+	clusterName := conn.ServerIdentity.ClusterName
+
+	// asyncEmitter makes sure that sessions do not block
+	// in case if connections are slow
+	asyncEmitter, err := process.newAsyncEmitter(conn.Client)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	streamer, err := events.NewCheckingStreamer(events.CheckingStreamerConfig{
+		Inner:       conn.Client,
+		Clock:       process.Clock,
+		ClusterName: clusterName,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	streamEmitter := &events.StreamerAndEmitter{
+		Emitter:  asyncEmitter,
+		Streamer: streamer,
+	}
```
Why do you need all of this stuff? This is for session recordings, right? Are there discovery sessions that we need to record?

r0mant replied:
I think this is used to emit audit events, but it probably can be simplified and we can just pass auth client instead of the stream emitter.

zmb3 replied:
So you just need an `Emitter`, not a `StreamerAndEmitter`

zmb3 commented on lib/service/service.go:
```diff
@@ -4562,9 +4595,9 @@ func (process *TeleportProcess) Close() error {
+	if !cfg.Auth.Enabled && !cfg.SSH.Enabled && !cfg.Proxy.Enabled && !cfg.Kube.Enabled && !cfg.Apps.Enabled && !cfg.Databases.Enabled && !cfg.WindowsDesktop.Enabled && !cfg.Discovery.Enabled {
 		return trace.BadParameter(
-			"config: enable at least one of auth_service, ssh_service, proxy_service, app_service, database_service, kubernetes_service or windows_desktop_service")
+			"config: enable at least one of auth_service, ssh_service, proxy_service, app_service, database_service, kubernetes_service, windows_desktop_service or discovery_service")
```
Discovery service seems more like an internal component than an agent like kube or database access. I'm not sure this change makes sense.

r0mant replied:
Replied above but yes, discovery service is intended to be a separate service/agent with these changes.

zmb3 commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,178 @@
+	AccessPoint auth.DiscoveryAccessPoint
+}
+
+// Server is a discovery server, It
```
It what?

zmb3 commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,178 @@
+// Server is a discovery server, It
+type Server struct {
+	*Config
+	events.StreamEmitter
```
Are you sure a stream emitter is necessary?

zmb3 commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+func (s *Server) Stop() {
+	if s.cloudWatcher != nil {
```
Under what circumstances can `cloudWatcher` be nil? I see a nil check here, but no nil checks on line 138 or 155.

zmb3 commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,157 @@
+		return trace.Wrap(err)
+	}
+	status := aws.StringValue(cmdOut.Status)
+	var code string
+	if status == ssm.CommandStatusSuccess {
+		code = libevents.SSMRunSuccessCode
+	} else {
+		code = libevents.SSMRunFailCode
+	}
```
```suggestion
	code := lib events.SSMRunFailCode
	if status == ssm.CommandStatusSuccess {
		code = libevents.SSMRunSuccessCode
	}
```

Commit:
Resolve comments

- use a regular events.Emitter
- resolve a thousand typos :)

r0mant commented on lib/cache/cache.go:
```diff
@@ -323,6 +324,18 @@ func ForWindowsDesktop(cfg Config) Config {
+	cfg.Watches = []types.WatchKind{
+		{Kind: types.KindCertAuthority, LoadSecrets: false},
+		{Kind: types.KindClusterName},
+		{Kind: types.KindNamespace, Name: apidefaults.Namespace},
```
Hmm, so shouldn't we include nodes here as well? Otherwise I think we may be going straight to auth server.

I think that we should include nodes here and below, where you initialize node watcher, it should be initialized with the access point (which is cache-based) instead of directly with the auth client.

r0mant commented on lib/cloud/clients.go:
```diff
@@ -455,6 +468,7 @@ type TestCloudClients struct {
 	AzurePostgres           azure.DBServersClient
 	AzurePostgresPerSub     map[string]azure.DBServersClient
 	AzureSubscriptionClient *azure.SubscriptionClient
+	SSM                     ssmiface.SSMAPI
```
Nit: Can you move `SSM` up so it's grouped together with other AWS API clients?

r0mant commented on lib/service/discovery.go:
```diff
@@ -0,0 +1,106 @@
+		ResourceWatcherConfig: services.ResourceWatcherConfig{
+			Component: teleport.ComponentDiscovery,
+			Log:       process.log.WithField(trace.Component, teleport.ComponentDiscovery),
+			Client:    conn.Client,
```
Use `accessPoint` here, not `conn.Client` (and add nodes to the cache like I mentioned above).

r0mant commented on lib/service/discovery.go:
```diff
@@ -0,0 +1,106 @@
+		Emitter:     asyncEmitter,
+		AccessPoint: accessPoint,
+	})
+
```
```suggestion
```

r0mant commented on lib/service/discovery.go:
```diff
@@ -0,0 +1,106 @@
+		log.Info("Exited.")
+	})
+
+	process.BroadcastEvent(Event{Name: DiscoveryReady, Payload: nil})
+	log.Infof("Discovery service has successfully started")
```
Should this be moved after you start the service below?

r0mant commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,174 @@
+		Config: cfg,
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component:       teleport.ComponentDiscovery,
+			trace.ComponentFields: logrus.Fields{},
```
```suggestion
```

r0mant commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,174 @@
+			}
+			s.log.Debugf("EC2 instances discovered (AccountID: %s, Instances: %v), starting installation", instances.AccountID, instIDs)
+			if err := s.handleInstances(&instances); err != nil {
+				if trace.IsNotFound(err) {
+					s.log.Debug("All discovered EC2 instances are already part of the cluster.")
```
It looks like `handleInstances` returns `nil` when all instances are already enrolled, not `NotFound` error.

r0mant commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,174 @@
+	}
+}
+
+func (s *Server) Start() error {
```
```suggestion
// Start starts the discovery service.
func (s *Server) Start() error {
```

r0mant commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,174 @@
+	return nil
+}
+
+func (s *Server) Stop() {
```
```suggestion
// Stop stops the discovery service.
func (s *Server) Stop() {
```

r0mant commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,174 @@
+			trace.Component:       teleport.ComponentDiscovery,
+			trace.ComponentFields: logrus.Fields{},
+		}),
+		ctx: ctx,
```
I would wrap the parent context in a local context with a cancel function so discovery service can be stopped independently of the outer process.

r0mant commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,174 @@
+}
+
+func (s *Server) Stop() {
+	s.cloudWatcher.Stop()
```
We should also call the cancel the discovery service's local context here (I left a comment above).

r0mant reviewed (commented):
I think we're close and I don't have a lot of outstanding comments anymore, this looks pretty good to me now bar a few additional suggestions. @tigrato @zmb3 - you requested changes here before, could you take another look?

@rosstimothy @fspmarshall We have implemented discovery service here, are you concerns addressed?

tigrato reviewed (commented):
Is there any process that prevents multiple SSM calls to the same instance?
I'm asking because I mentioned this earlier and something must have slipped my mind while reviewing it.

My main problem is that if the SSM + node join takes longer than the discovery ticker, we can run the same script multiple times against the same machine.

I would also change the configuration so that it is possible to extend the service to other resources like dbs, kubernetes... but it probably requires some input from @r0mant

Commit:
Resolve comments

tigrato commented:
> > Is there any process that prevents multiple SSM calls to the same instance? I'm asking because I mentioned this earlier and something must have slipped my mind while reviewing it.
> > My main problem is that if the SSM + node join takes longer than the discovery ticker, we can run the same script multiple times against the same machine.
> > I would also change the configuration so that it is possible to extend the service to other resources like dbs, kubernetes... but it probably requires some input from @r0mant
> 
> The script itself includes a file lock that is kept for the duration of the run, so it can only be run once at a time (ssm command executions do appear to be free), so if the script is being run on a node already it should just exit, there was some discussion of using a semaphore to lock runs however that may just be a followup patch

Uhmn, and if the script didn't start yet because SSM is slow or down?
SSM is part of the automation section for AWS, right? If yes, as far as I know, it is paid after the free tier limit (100k requests or 5k seconds of aggregated running time): https://aws.amazon.com/systems-manager/pricing/#Automation

Commit:
resolve comments, fix a bad merge

tigrato commented on lib/config/configuration.go:
```diff
@@ -1060,8 +1064,13 @@ func applySSHConfig(fc *FileConfig, cfg *service.Config) (err error) {
+
+func applyDiscoveryConfig(fc *FileConfig, cfg *service.Config) {
+	cfg.Discovery.Enabled = fc.Discovery.Enabled()
+	for _, matcher := range fc.Discovery.AWSMatchers {
```
Yeah, probably. I didn't think about it

Commit:
Fail when a non ec2 matcher type is configured

Commit:
fix lint-go

r0mant commented on lib/config/fileconf.go:
```diff
@@ -438,9 +442,15 @@ func (conf *FileConfig) CheckAndSetDefaults() error {
+
+	for _, matcher := range conf.Discovery.AWSMatchers {
+		for _, serviceType := range matcher.Matcher.Types {
+			if serviceType != "ec2" {
```
Nit: Can we move "ec2" into a constant and have a global variable somewhere e.g. `supportedDiscoveryServices` that will make it easier to extend this?

And also tweak the message below e.g. "discovery service type does not support %q, supported resource types are: %v" like we do elsewhere.

r0mant commented on lib/config/fileconf.go:
```diff
@@ -1108,6 +1109,13 @@ func (ssh *SSH) X11ServerConfig() (*x11.ServerConfig, error) {
+	Service `yaml:",inline"`
+
+	// AWSMatchers are used to match EC2 instances
+	AWSMatchers []AWSEC2Matcher `yaml:"aws,omitempty"`
```
As discussed offline, the current structure is ok since we're using "types" field to filter by types.

However, since we'll be using this section for other resource types as well, not just EC2 instances, I don't think `AWSEC2Matcher` name is really fitting. Can we rename it to just `AWSMatcher`?

r0mant commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,192 @@
+	"github.com/sirupsen/logrus"
+)
+
+type Config struct {
+	Clients     cloud.Clients
+	Matchers    []services.AWSMatcher
+	NodeWatcher *services.NodeWatcher
+	Emitter     apievents.Emitter
+	AccessPoint auth.DiscoveryAccessPoint
+}
```
Nit: Add godocs on the struct and its public fields.

r0mant commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,192 @@
+	cancelfn context.CancelFunc
+}
+
+func New(ctx context.Context, cfg *Config) (*Server, error) {
```
Nit: Add godoc.

r0mant commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,192 @@
+		logInstances.WriteString(aws.StringValue(inst.InstanceId) + ", ")
+	}
+	if len(instances.Instances) > 10 {
+		logInstances.WriteString(fmt.Sprintf("... + %d instance IDs trunksated", len(instances.Instances)-10))
```
```suggestion
		logInstances.WriteString(fmt.Sprintf("... + %d instance IDs truncated", len(instances.Instances)-10))
```

r0mant commented on lib/srv/discovery/discovery.go:
```diff
@@ -0,0 +1,192 @@
+	}
+	s.filterExistingNodes(instances)
+	if len(instances.Instances) == 0 {
+		return trace.NotFound("all fetched nodes already enrolled.")
```
```suggestion
		return trace.NotFound("all fetched nodes already enrolled")
```

r0mant commented on lib/srv/server/ssm_install.go:
```diff
@@ -0,0 +1,156 @@
+	"context"
+	"errors"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ssm"
+	"github.com/aws/aws-sdk-go/service/ssm/ssmiface"
```
Looks like we use a mix of v1 and v2 SDK's in the code. I agree it'd be better to eventually migrate to v2 for everything but since we already started with v1 and it looks like v2 is not a drop-in replacement, I think it's ok for now, we can refactor later.

r0mant reviewed (approved):
I didn't find any blockers anymore, this lgtm now. @tigrato @rosstimothy Any other feedback guys or can we ship it?

@lxea Since we introduced a new service here, before we merge this let's please test a couple of things:

- [x] [CA rotation](https://goteleport.com/docs/management/operations/ca-rotation/) when discovery service is running standalone or together with, say, SSH service. It should not be affected but just in case.
- [x] [Graceful restarts](https://goteleport.com/docs/management/admin/daemon/#step-33-restart-the-teleport-daemon) to make sure that discovery service shuts down cleanly and does not block the process from shutting down.

Commit:
Resolve comments

rosstimothy reviewed (commented):
Thanks for splitting this out into its own service @lxea! Looks good to me from a scale perspective. I think that `discovery.Server` could benefit from some tests, it currently has none and there are no integration tests for the discovery service making it very likely things could get broken as more resources are migrated into it.

Commit:
Resolve comments, add initial test (currently broken)

Commit:
Fix log string so only 1 pair of [] are used

Commit:
Chunk instances for sending commands

Commit:
add 'isInitialized' to watchers

Commit:
Add test for chunked discovery, log output

Commit:
lints

r0mant commented:
@lxea Is this ready to be merged?

Commit:
explicetly set matcher.Tags to "*":"*" if its unset
